File: app/Packages/ONTFeatures/Sources/SearchFeature/Application/SearchModel.swift
import Foundation
import ONTKit
import Observation

/// Le modèle de la recherche.
@MainActor
@Observable
public final class SearchModel {
    private let index: any SearchIndex
    private let glossary: any GlossaryRepository
    private let corpus: any CorpusRepository

    public var query = "" { didSet { run() } }
    public var scope = SearchScope.body { didSet { run() } }
    public private(set) var hits: [SearchHit] = []

    /// Les lemmes du glossaire — pour que taper « chesed » trouve aussi les
    /// passages où le terme ne paraît qu'en hébreu. Calculés une fois : le
    /// macro `@Observable` n'admet pas `lazy`.
    @ObservationIgnored private let lemmas: Set<String>

    public init(
        index: any SearchIndex,
        glossary: any GlossaryRepository,
        corpus: any CorpusRepository
    ) {
        self.index = index
        self.glossary = glossary
        self.corpus = corpus
        self.lemmas = Set(((try? glossary.entries()) ?? []).map(\.lemma))
    }

    private func run() {
        hits = SearchEngine.search(query, in: index.records(), scope: scope, lemmas: lemmas)
    }

    /// Le titre d'un livre, pour l'affichage d'un résultat.
    public func bookTitle(_ bookId: String) -> String {
        corpus.allBooks().first { $0.id == bookId }?.title ?? bookId
    }
}

File: android/ontfeatures/src/main/kotlin/com/labibleont/ont/features/search/SearchModel.kt
package com.labibleont.ont.features.search

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.labibleont.ont.kit.ports.GlossaryRepository
import com.labibleont.ont.kit.ports.SearchIndex
import com.labibleont.ont.kit.search.SearchEngine
import com.labibleont.ont.kit.search.SearchHit
import com.labibleont.ont.kit.search.SearchScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * La recherche dans le corpus.
 *
 * ## Chargé à la première requête, pas au lancement
 *
 * `search.json` fait 640 Ko. Le décoder au démarrage retarderait l'ouverture
 * pour une fonction qu'on n'emploie pas à chaque session.
 *
 * ## Le délai avant de chercher n'est pas un confort
 *
 * Sans lui, taper « chesed » lance six balayages du corpus, dont cinq dont on
 * jette le résultat. À l'échelle actuelle chacun prend quelques millisecondes ;
 * à soixante-dix livres rédigés, ils se verront. Le poser maintenant coûte
 * quatre lignes, le poser plus tard demandera de comprendre pourquoi la saisie
 * saccade.
 */
public class SearchModel(
    private val index: SearchIndex,
    private val glossaire: GlossaryRepository,
) : ViewModel() {

    public var requete: String by mutableStateOf("")
        private set

    public var portee: SearchScope by mutableStateOf(SearchScope.ALL)
        private set

    public var resultats: kotlin.collections.List<SearchHit> by mutableStateOf(emptyList())
        private set

    public var cherche: Boolean by mutableStateOf(false)
        private set

    private var lemmes: Set<String> = emptySet()
    private var enCours: Job? = null

    public fun saisir(texte: String) {
        requete = texte
        relancer()
    }

    public fun changerPortee(nouvelle: SearchScope) {
        portee = nouvelle
        relancer()
    }

    private fun relancer() {
        enCours?.cancel()
        val q = requete.trim()
        if (q.length < 2) {
            resultats = emptyList()
            cherche = false
            return
        }
        cherche = true
        enCours = viewModelScope.launch {
            delay(220)
            val trouves = withContext(Dispatchers.IO) {
                if (lemmes.isEmpty()) {
                    // Les lemmes rattrapent les passages où le terme ne paraît
                    // qu'en hébreu : taper « chesed » doit les trouver aussi.
                    lemmes = glossaire.entries().map { it.lemma }.toSet()
                }
                SearchEngine.search(q, index.records(), portee, lemmes)
            }
            resultats = trouves
            cherche = false
        }
    }
}
